### PR TITLE
Update torch.cat decomp for 0-dim

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130763

Fix for https://github.com/pytorch/pytorch/issues/130615

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Review by mlazos — APPROVED** [review]

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -300,7 +300,13 @@ def non_empty_tensor(x):
         # runtime assert forcing u0 to be zero.  So if this hasn't happened,
         # we know that the unbacked SymInt has appropriate size and there are
         # no problems.
-        return len(x.shape) != 1 or guard_size_oblivious(x.shape[0] > 0)
+        if len(x.shape) == 1 and guard_size_oblivious(x.shape[0] == 0):
+            return False
+
+        if dim < len(x.shape) and guard_size_oblivious(x.shape[dim] == 0):
```
Comment (peterbell10) [inline torch/_inductor/decomposition.py:306]: `guard_size_oblivious` implies an unbacked symint could get through this with a runtime value of 0, right. Would we still IMA in that case?

Comment (eellison) [reply torch/_inductor/decomposition.py:306]: Yea, it's a good point, will do in follow up (we would still want this change anyway)

Comment (eellison) [reply torch/_inductor/decomposition.py:306]: The issue on master is we are ignoring the masked load and turning into a load. when I set [needs_where=True](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L1031) the IMA goes away. cc @isuruf, coming from your chagne

Comment (isuruf) [reply torch/_inductor/decomposition.py:306]: @eellison, I tried `need_where = True` on the test case and it still failed with a IMA before this PR was changed. Was there any other change that you made?

Comment (eellison) [reply torch/_inductor/decomposition.py:306]: Sorry, you're right just retested and it imas even with the revert my bad. However, the unbacked case also still works

```
import torch
import torch.nn as nn


def fn(out, x):
    return torch.cat([x, out[0]], dim = 0)

opt_fn = torch.compile(fn, dynamic=True)
out = torch.randn(1, 0, 64, device="cuda")
torch._dynamo.decorators.mark_unbacked(out, 1)
x = torch.randn(128, 64, device="cuda")

fn(out, x)
opt_fn(out, x)
```

Doesn't we can hit this ima on master although i do think there's still something missing with current masked load
